src/components/lab/NumberInput/NumberInput.tsx

```diff
@@ -0,0 +1,326 @@
+
+export const NumberInput = React.forwardRef<HTMLSpanElement, NumberInputProps>(function NumberInput(
+    {endContent, defaultValue: externalDefaultValue, ...props},
+    ref,
+) {
+    const {
+        value: externalValue,
+        onChange: handleChange,
+        onUpdate: externalOnUpdate,
+        min: externalMin,
+        max: externalMax,
+        shiftMultiplier: externalShiftMultiplier = 10,
+        step: externalStep = 1,
+        size = 'm',
+        view = 'normal',
+        disabled,
+        hiddenControls,
+        validationState,
+        onFocus,
+        onBlur,
+        onKeyDown,
+        allowMouseWheel = false,
+        allowDecimal = false,
+        className,
+    } = props;
+
+    const {
+        min,
+        max,
+        step: baseStep,
+        value: internalValue,
+        defaultValue,
+        shiftMultiplier,
+    } = getInternalState({
+        min: externalMin,
+        max: externalMax,
+        step: externalStep,
+        shiftMultiplier: externalShiftMultiplier,
+        allowDecimal,
+        value: externalValue,
+        defaultValue: externalDefaultValue,
+    });
+
+    const [value, setValue] = useControlledState(
+        internalValue,
+        defaultValue ?? null,
+        externalOnUpdate,
+    );
+
+    const [inputValue, setInputValue] = React.useState(getStringValue(value));
+
+    React.useEffect(() => {
+        const stringPropsValue = getStringValue(value);
+        setInputValue((currentInputValue) => {
+            if (!areStringRepresentationOfNumbersEqual(currentInputValue, stringPropsValue)) {
+                return stringPropsValue;
+            }
+            return currentInputValue;
+        });
+    }, [value]);
+
+    const clamp = true;
+
+    const [active, setActive] = React.useState(false);
+    const safeValue = value ?? 0;
+
+    const state = getInputControlState(validationState);
+
+    const canIncrementNumber = safeValue < (max ?? Number.MAX_SAFE_INTEGER);
+
+    const canDecrementNumber = safeValue > (min ?? Number.MIN_SAFE_INTEGER);
+
+    const innerControlRef = React.useRef<HTMLInputElement>(null);
+    const fieldRef = useFormResetHandler({
+        initialValue: value,
+        onReset: setValue,
+    });
+    const handleRef = useForkRef(props.controlRef, innerControlRef, fieldRef);
+
+    const handleIncrement = (
+        e:
+            | React.MouseEvent<HTMLButtonElement>
+            | React.WheelEvent<HTMLInputElement>
+            | React.KeyboardEvent<HTMLInputElement>,
+    ) => {
+        const step = e.shiftKey ? shiftMultiplier * baseStep : baseStep;
+        if (canIncrementNumber) {
+            const newValue = clampToNearestStepValue({
+                value: safeValue + step,
+                step: baseStep,
+                min,
+                max,
+                direction: 'up',
+            });
+            setValue?.(newValue);
+            setInputValue(newValue.toString());
+        }
+    };
+
+    const handleDecrement = (
```

Comment:
handleIncrement and handleDecrement look very alike. Let's combine them into one function with the signature `handleValueDelta(delta: number)`, and use like `handleValueDelta(1)`, `handleValueDelta(-10)` (if shift is pressed, and derection is negative).